introduced iconset infrastructure

kaikreuzer opened the pull request:
Signed-off-by: Kai Kreuzer <kai@openhab.org>

kaikreuzer commented:
This PR is just as a preview to give you the chance to provide feedback.
I still have to add some icons, tests and documentation.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.classic/snippets/rollerblind.html:
```diff
@@ -1,5 +1,5 @@
 <li>
-	<img src="../icon/%icon%.png" width=29 height=29 class="iFull" />
+	<img src="../icon/%icon%?format=%format%&state=%state%" width=29 height=29 class="iFull" />
```
%icon% is always the category of an item, right? Should we change it to %category% to make it clear

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.classic/src/main/java/org/eclipse/smarthome/ui/classic/internal/render/AbstractWidgetRenderer.java:
```diff
@@ -158,4 +164,37 @@ protected String processColor(Widget w, String snippet) {
+        return "NULL";
+    }
+
+    protected String getIcon(Widget w) {
```
Same question here: What does it actually return? a category, an Icon (id)? I think the concrete item is at the end calculated out of the category, the pattern and the state. So "getIcon" might be a bit misleading.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/AbstractResourceIconProvider.java:
```diff
@@ -0,0 +1,110 @@
+ * If a specific resource for a state is available, it will be used. If not, the default icon without a state postfix is
+ * used. If the state is a decimal number between 0 and 100, the implementation will look for a resource with the next
+ * smaller state postfix available. Example: For category "DimmableLight" and state 84, it will check for the resources
+ * dimmablelight-82.png, dimmablelight-81.png, dimmablelight-80.png and return the first one it can find.
```
This sounds like it performs really bad. Does it tries to look up all the files? Will check implementation.

kaikreuzer replied:
Note that this is only done once - on every subsequent request a 304 is sent: https://github.com/eclipse/smarthome/pull/445/files#diff-27e9afb0b912dc64ed3e95af0adcfcbdR101

kaikreuzer replied:
And as discussed, the current implementation is already doing the same thing - just within the ItemUIRegistry, not in the servlet.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/AbstractResourceIconProvider.java:
```diff
@@ -0,0 +1,110 @@
+                if (stateAsDouble > 0 && stateAsDouble < 100) {
+                    for (int i = stateAsDouble.intValue(); i >= 0; i--) {
+                        String resourceWithNumberState = category + "-" + i + "." + format.toString().toLowerCase();
+                        if (hasResource(iconSetId, resourceWithNumberState)) {
```
In worst case if the state is 99 it tries to look up 99 files? And why looking for the smaller resource? If there are Icons for 0, 25, 75 and 100, wouldn´t you expect that it returns the Icon for 75 if the status is 74?

kaikreuzer replied:
Please see https://github.com/eclipse/smarthome/pull/445/files#diff-5d3a30bfebd36343f1f1d61e9f589de1L462.
This is the existing logic, I didn't change anything about it so that it does not cause any regression.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/AbstractResourceIconProvider.java:
```diff
@@ -0,0 +1,110 @@
+     * @param resourceName the name of the resource
+     * @return the content as a stream or null, if the resource does not exist
+     */
+    abstract protected InputStream getResource(String iconSetId, String resourceName);
```
For another API (Sound resource API) I just searched for a good and abstract return value similar to InputStream, but which does not open the stream. Guava has a good class for it: ByteSource (http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/io/ByteSource.html).

It has a lot of helper methods and serves as a good return type if you want to provide access to raw data without giving around opened streams. But unfortunately our Guava version is too old (we have 10.0.1 but we need 14.0).

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/IconProvider.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @return a set of icon sets in the default locale
+     */
+    Set<IconSet> getIconSets();
```
Does one IconProvider really needs to return multiple IconSets?

Otherwise we could name the interface IconSetProvider.

kaikreuzer replied:
The CustomIconProvider can serve icons for different iconsets as it just reads them from the file system from different folders.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/IconProvider.java:
```diff
@@ -0,0 +1,67 @@
+     * @return an Integer value defining the priority (higher is more important) or <code>null</code>, if this provider
+     *         cannot deliver an icon
+     */
+    Integer hasIcon(String category, String iconSetId, Format format);
```
Let´s say I want to implement an IconProvider. What priority do I usually have to return? 0, 1? Does it has to be higher than some defined value? Like the priority of the default icons?

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/IconProvider.java:
```diff
@@ -0,0 +1,67 @@
+     * @param format the format of the stream (usually either png or svg)
+     * @return a byte stream of the icon in the given format
+     */
+    InputStream getIcon(String category, String iconSetId, String state, Format format);
```
Is this method only called, if "hasIcon" has returned a priority?

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/IconSet.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.Set;
+
+/**
+ * This is a pojo that provides some meta-information about available icon sets.
```
I wouldn´t mention he term "pojo" in javadoc.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/CustomIconProvider.java:
```diff
@@ -45,19 +53,19 @@ public InputStream getIcon(String iconName) {
-        return null;
+    @Override
+    public Set<IconSet> getIconSets(Locale locale) {
+        return Collections.emptySet();
```
Why emptySet? Could´nt we iterate over the folders to get the IconSets?

kaikreuzer replied:
You would not have any meta-info from the folders and the custom provider is only meant to augment existing icon sets by custom icons.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -97,21 +102,62 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
-
-        String iconName = StringUtils.substringBeforeLast(filename, ".");
+        String category = getCategory(req);
+        Format format = getFormat(req);
```
if the UI sends the format, how does it now if there are icons with the according format?

kaikreuzer replied:
From the iconset metadata.

dnobel replied:
OK, that means if an icon set wants to support "svg" all icons has be SVG, right? But if I want to override it with a custom Icon, does it have to be SVG, too? From the UI I don´t know if it will return a custom icon or not, so I will always request it with svg.

kaikreuzer replied:
Yes, this is my assumption. If the icon set meta-data claims that it supports SVG, this should be true for all icons (even the custom ones).

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -97,21 +102,62 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+            ServletOutputStream os = resp.getOutputStream();
+            InputStream is = topProvider.getIcon(category, iconSetId, state, format);
+            IOUtils.copy(is, os);
+            is.close();
```
Does this has to be moved into a finally block?

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -97,21 +102,62 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+    }
+
+    private String getCategory(HttpServletRequest req) {
+        String filename = StringUtils.substringAfterLast(req.getRequestURI(), "/");
```
category?

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java:
```diff
@@ -97,21 +102,62 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+
+    private String getCategory(HttpServletRequest req) {
+        String filename = StringUtils.substringAfterLast(req.getRequestURI(), "/");
+        return StringUtils.substringBeforeLast(filename, ".");
```
why is there a "dot" in the path?

kaikreuzer replied:
This is for backward compatibility, so that clients that still send a /icon/switch.png still work.

dnobel commented on extensions/ui/iconset/org.eclipse.smarthome.ui.iconset.classic/src/main/java/org/eclipse/smarthome/ui/iconset/classic/internal/ClassicIconProvider.java:
```diff
@@ -0,0 +1,78 @@
+        formats.add(Format.PNG);
+        formats.add(Format.SVG);
+        IconSet iconSet = new IconSet(ICONSET_ID, "Classic Icons",
+                "This is a modernized version of the original icon set of openHAB 1.", formats);
```
Use i18n Provider to get english texts out of a properties file

dnobel commented on extensions/ui/iconset/org.eclipse.smarthome.ui.iconset.classic/src/main/java/org/eclipse/smarthome/ui/iconset/classic/internal/ClassicIconProvider.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected InputStream getResource(String iconSetId, String resourceName) {
+        if (ICONSET_ID.equals(iconSetId)) {
```
Is it possible that someone call getResource for an IconSet that is not provided by this this IconProvider?

kaikreuzer replied:
You never know...

dnobel commented on extensions/ui/iconset/org.eclipse.smarthome.ui.iconset.classic/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,11 @@
+Manifest-Version: 1.0
```
Why do you need an own bundle for the ClassicUI Icon set? Can´t it be put into the ClassicUI bundle? And does it really contain 4 Icons only?

kaikreuzer replied:
> Why do you need an own bundle for the ClassicUI Icon set?

Because it is not the ClassicUI icon set, but the Classic icon set. It is used by the Android app, the iOS app, the Basic UI and soon also the Paper UI as well :-)

> And does it really contain 4 Icons only?

This is the status quo, see https://github.com/eclipse/smarthome/tree/master/distribution/smarthome/runtime/icons.
But once this PR is merged, I plan to have the new openHAB 2 icon set contributed to ESH, see https://github.com/openhab/openhab2/pull/291#issuecomment-119709130.

dnobel commented:
Some general comments/questions:

* I´m not sure about the performance. If you have a lot of IconSets for each icon request all ItemProviders are asked if they provide an item for the category, pattern and state combination. This results in a lot of file lookups, especially for dimmer items. But I just read your comments, if this does not make any performance problems, OK.
* In the sitemap you explicitly set an "icon", not a category. But at the end it will be treated as a category, when given to the IconProvider. This seems to be a bit inconsistent. Even if it might not be relevant for the current sitemap, it´s something that must be discussed for the new sitemap concept.
* The format: If a UI always requests a SVG, because the chosen Icon Set supports SVG, someone who wants to add his own icon has also provide a SVG version of this icon. Because I think from the current implementation the file lookup will fail and no icon will be provided.
* I don´t really understand the Icon Set ID. Let´s say we have different Icon Sets availalbe, does the UI always have to send an Icon Set ID, because otherwise no icon will be returned? Should we store a mapping between Icon Set IDs and IconProviders, to not always iterate through all IconProviders? If I specifiy a concrete Icon Set, can other Icon Providers still return Icons, even if they don´t support the requested set?

kaikreuzer commented:
Thanks a lot for the review. I have incorporated all your feedback.

> I´m not sure about the performance.

I don't think this is critical. Most of the behavior actually existed already before. I doubt that you will have many (>5) icon providers ever in the system. And by returning 304s, all efforts are usually just a one time effort.

> In the sitemap you explicitly set an "icon", not a category.

I have updated all code to now speak about categories, I only refrained from updating the DSL, since this would have too much impact. But yes, for a future sitemap version, this wording should then be aligned.

> The format: If a UI always requests a SVG

As written in my other comment, I think the assumption is fine that if an iconset claims to support SVG, the UIs are allowed to request it for all icons. Anyhow, it can also always happen that a certain category is not available, independent from the format. We could discuss, whether the icon servlet should in such cases never return 404, but rather deliver an empty image to the UI, so that this does not need to deal with 404s.

> does the UI always have to send an Icon Set ID, because otherwise no icon will be returned?

No, the icon servlet uses "classic" as a hardcoded fallback, but with https://github.com/kaikreuzer/smarthome/commit/18bbb9c326e6e2be208fc012b9e08268859e0114, this is now also configurable through ConfigAdmin.
Additionally, the UIs could offer a configuration which iconset should be used - for the Classic/Basic UI, this will e.g. also be configured through the ConfigAdmin.

> can other Icon Providers still return Icons, even if they don´t support the requested set?

Yes, as there is no notion of "supports a set". As you can see with the CustomIconProvider, it is absolultly fine to provide icons even though the provider does not declare any sets themselves.

kaikreuzer commented:
I have added documentation and a skeleton for the tests (still to be done...)

kaikreuzer commented:
Finished with the implementation of the tests. @dnobel, please check - if all is ok, I will squash the commits before the merge.

dnobel commented on bundles/ui/org.eclipse.smarthome.ui.icon.test/src/test/groovy/org/eclipse/smarthome/ui/icon/test/IconServletTest.groovy:
```diff
@@ -0,0 +1,112 @@
+
+    @Test
+    void testOldUrlStyle() {
+        calledProvider = null
```
I would set this in setUp and/or tearDown method to null

dnobel commented on docs/documentation/features/ui/icons.md:
```diff
@@ -0,0 +1,30 @@
+
+Eclipse SmartHome comes with a flexible infrastructure for handling icons that are to be used within user interfaces.
+This bunde registers a servlet under the url `/icon/`, which can then be easily queried for icons using GET requests of the form `/icon/<category>?state=<state>&format=[png|svg]&iconset=<iconsetid>`, where 
+- `category` is one from the [list of channel categories](../../development/bindings/thing-definition.html#channel-categories)
```
maybe you can mention that it can also be a custom category for custom icons?

dnobel commented on docs/pom.xml:
```diff
@@ -87,6 +106,36 @@
+                        </goals>
+                        <configuration>
+                            <source>
+                                def iconsets = new File(project.basedir, 'documentation/features/ui/iconset')
```
In the Swagger PR Dancho moved that out to own groovy files for better maintenance. I think we should also do it here

kaikreuzer commented:
@dnobel, I have incorporated your feedback and squashed my commits.

dnobel commented:
I guess that means I should merge it? :)
